sway/desktop/transaction.c

```diff
@@ -0,0 +1,398 @@
+					instruction->state.view_width,
+					instruction->state.view_height);
+			++transaction->num_waiting;
+		}
+		list_add(con->instructions, instruction);
+	}
+	transaction->num_configures = transaction->num_waiting;
+	if (server.debug_txn_timings) {
+		clock_gettime(CLOCK_MONOTONIC, &transaction->commit_time);
+	}
+	if (server.head_transaction) {
+		// There is another transaction in progress - we must add this one to
+		// the queue so we complete after it.
+		struct sway_transaction *tail = server.head_transaction;
+		while (tail->next) {
+			tail = tail->next;
+		}
+		tail->next = transaction;
+	} else if (transaction->num_waiting) {
+		// There are no other transactions, but we're not applying immediately
+		// so we must jump in the queue so others will queue behind us.
+		server.head_transaction = transaction;
+	} else {
+		// There are no other transactions in progress, and this one has nothing
+		// to wait for, so we can skip the queue.
+		wlr_log(L_DEBUG, "Transaction %p has nothing to wait for", transaction);
+		transaction_apply(transaction);
+		transaction_destroy(transaction);
+		return;
+	}
+
+	if (transaction->num_waiting) {
+		// Set up a timer which the views must respond within
+		transaction->timer = wl_event_loop_add_timer(server.wl_event_loop,
+				handle_timeout, transaction);
+		wl_event_source_timer_update(transaction->timer, TIMEOUT_MS);
+	}
+
+	// The debug tree shows the pending/live tree. Here is a good place to
+	// update it, because we make a transaction every time we change the pending
+	// tree.
+	update_debug_tree();
+}
+
+static void set_instruction_ready(
+		struct sway_transaction_instruction *instruction) {
+	instruction->ready = true;
+	struct sway_transaction *transaction = instruction->transaction;
+
+	if (server.debug_txn_timings) {
+		struct timespec now;
+		clock_gettime(CLOCK_MONOTONIC, &now);
+		struct timespec *start = &transaction->commit_time;
+		float ms = (now.tv_sec - start->tv_sec) * 1000 +
+			(now.tv_nsec - start->tv_nsec) / 1000000.0;
+		wlr_log(L_DEBUG, "Transaction %p: %li/%li ready in %.1fms (%s)",
+				transaction,
+				transaction->num_configures - transaction->num_waiting + 1,
+				transaction->num_configures, ms,
+				instruction->container->name);
+
+	}
+
+	// If all views are ready, apply the transaction.
+	// If the transaction has timed out then its num_waiting will be 0 already.
+	if (transaction->num_waiting > 0 && --transaction->num_waiting == 0) {
+#if !TRANSACTION_DEBUG
+		wlr_log(L_DEBUG, "Transaction %p is ready", transaction);
+		wl_event_source_timer_update(transaction->timer, 0);
+		transaction_progress_queue();
+#endif
+	}
+}
+
+void transaction_notify_view_ready(struct sway_view *view, uint32_t serial) {
+	for (int i = 0; i < view->swayc->instructions->length; ++i) {
+		struct sway_transaction_instruction *instruction =
+			view->swayc->instructions->items[i];
+		if (instruction->serial == serial && !instruction->ready) {
+			set_instruction_ready(instruction);
```

Comment:
A client can receive configure n and n+1 and only ACK n+1. That means we should `set_instruction_ready` for all instructions that have a serial <= ACKed serial.

Reply:
Thinking about this more, a much more simple option would be to mark skipped instructions as ready as you suggested, but not apply any transactions at all unless the transaction queue can be completely processed.

Reply:
Possibly far fetched, but would it be possible to have transaction 1 with views A & B, transaction 2 with views B & C.
With what you just said, if A acks 1st configure, B acks 2nd configure directly, when C is done we'll redraw B & C but not A?

I like your previous idea (mark first transaction as waiting-for-2nd, then apply both at once when 2nd is ready) better

Reply:
> if A acks 1st configure, B acks 2nd configure directly, when C is done we'll redraw B & C but not A

When would you redraw A? (or more correctly, when would you apply A's state?)

I've changed the implementation to one which waits for the entire queue to be ready before applying them. This is easier to implement than trying to group them based on dependencies, and should have the same result as long as there's 2 or less transactions in progress at the same time. And even if you have 3 transactions, you'd have to have 2 of them in separate parts of the tree before the dependency-group option has a benefit over the wait-for-full-queue option.

Reply:
Yeah, I think that'll work. It's less subtle than what was suggested earlier but it should be OK in practice.